proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2296,478 @@ crunches doIPCTransfer, possibleSwitchTo
+                                       valid_release_queue' and valid_queues and valid_objs' and
+                                       valid_queues' and (\<lambda>s. data \<noteq> None \<longrightarrow> reply_at' (the data) s \<and>
+                                       replySCs_of s (the data) = None)"
+                                       in hoare_strengthen_post[rotated])
+                 apply (fastforce split: option.splits)
+                apply (wpsimp wp: hoare_vcg_imp_lift)
+               apply (assumption)
+              apply (prop_tac "(tcb_at' t and tcb_at' a and valid_pspace' and cur_tcb' and
+                                ep_at' ep and valid_release_queue and valid_release_queue' and
+                                valid_queues and valid_objs' and valid_mdb' and valid_queues') s'",
+                     assumption)
+    apply clarsimp
+    apply (case_tac data; clarsimp simp: receive_reply_def)
+    apply (subgoal_tac "reply_at' aa s'", simp)
+    apply (frule (1) replySCs_of_cross, simp)
+    apply (erule cross_relF[OF _ reply_at'_cross_rel])
+    apply (clarsimp simp: obj_at_def reply_sc_reply_at_def is_reply)
+               apply (wpsimp wp: gts_wp)
+              apply (wpsimp wp: gts_wp')
+             apply (clarsimp simp: ep_relation_def split: list.splits)
+            apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift' hoare_vcg_disj_lift)
+           apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift' hoare_vcg_disj_lift)
+          \<comment> \<open>end of main Hoare triples\<close>
+          apply (subgoal_tac "tcb_at a s")
+          apply (subgoal_tac "a \<noteq> ep")
+           apply (clarsimp simp: invs_def valid_state_def valid_pspace_def valid_sched_def
+                                 valid_sched_action_def pred_tcb_at_eq_commute)
+           apply (prop_tac "(a, EPRecv) \<in> state_refs_of s ep")
+            apply (clarsimp simp: state_refs_of_def obj_at_def)
+           apply (frule (1) sym_refsD, simp)
+           apply (frule TCBBlockedRecv_in_state_refs_of)
+           apply (clarsimp simp: invs_def pred_tcb_at_eq_commute cong: conj_cong)
+            apply (intro conjI impI)
+                 apply (clarsimp simp: valid_ep_def split: list.splits)
+                apply (erule (1) if_live_then_nonz_capD)
+                apply (clarsimp simp: obj_at_def live_def)
+               apply (erule weak_valid_sched_action_scheduler_action_not)
+               apply (clarsimp simp: obj_at_def pred_tcb_at_def)
+              apply (clarsimp, erule (1) FalseI[OF idle_no_ex_cap], clarsimp simp: valid_idle_def)
+        apply (case_tac "receive_reply x"; simp)
+        apply (subgoal_tac "data = Some aa", simp)
+         apply (subgoal_tac "reply_tcb_reply_at ((=) (Some a)) aa s", simp)
+         apply (subgoal_tac "reply_sc_reply_at (\<lambda>a. a = None) aa s", simp)
+          apply (intro conjI)
+                 apply (clarsimp simp: sk_obj_at_pred_def obj_at_def)
```

Comment:
I think some of thse subgoals could be `fastforce`d, which could make it a little easier to read? No big deal, though